// @lwc/synthetic-shadow/lib/polyfills/iframe-content-window/patch.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
const language_1 = require("../../shared/language");
const node_1 = require("../../faux-shadow/node");
function apply() {
    // the iframe property descriptor for `contentWindow` should always be available, otherwise this method should never be called
    const desc = language_1.getOwnPropertyDescriptor(HTMLIFrameElement.prototype, 'contentWindow');
    const { get: originalGetter } = desc;
    desc.get = function () {
        const original = originalGetter.call(this);
        if (language_1.isNull(original) || !node_1.isNodeShadowed(this)) {
            return original;
        }
        // only if the element is an iframe inside a shadowRoot, we care about this problem
        // because in that case, the code that is accessing the iframe, is very likely code
        // compiled with proxy-compat transformation. It is true that other code without those
        // transformations might also access an iframe from within a shadowRoot, but in that,
        // case, which is more rare, we still return the wrapper, and it should work the same,
        // this part is just an optimization.
        return wrapIframeWindow(original);
    };
    language_1.defineProperty(HTMLIFrameElement.prototype, 'contentWindow', desc);
}
exports.default = apply;
function wrapIframeWindow(win) {
    return {
        postMessage() {
            // Typescript does not like it when you treat the `arguments` object as an array
            // @ts-ignore type-mismatch
            return win.postMessage.apply(win, arguments);
        },
        blur() {
            // Typescript does not like it when you treat the `arguments` object as an array
            // @ts-ignore type-mismatch
            return win.blur.apply(win, arguments);
        },
        close() {
            // Typescript does not like it when you treat the `arguments` object as an array
            // @ts-ignore type-mismatch
            return win.close.apply(win, arguments);
        },
        focus() {
            // Typescript does not like it when you treat the `arguments` object as an array
            // @ts-ignore type-mismatch
            return win.focus.apply(win, arguments);
        },
        get closed() {
            return win.closed;
        },
        get frames() {
            return win.frames;
        },
        get length() {
            return win.length;
        },
        get location() {
            return win.location;
        },
        set location(value) {
            win.location = value;
        },
        get opener() {
            return win.opener;
        },
        get parent() {
            return win.parent;
        },
        get self() {
            return win.self;
        },
        get top() {
            return win.top;
        },
        get window() {
            return win.window;
        },
    }; // this is limited
}
